### ISSUE TITLE
Chat participant API

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode#199908]
We have finalized the Language Model and Chat Participant API 🚀 
Thus, you can now publish your extensions to the VS Marketplace. This API is finalized in VS Code Insiders, and will be finalized in VS Code Stable in July. In your package.json, you should set

```
"engines": {
	"vscode": "^1.90.0"
},
```

and VS Code Stable will gracefully handle your extensions and ignore your usage of Language Model and Chat API until the API gets finalized in Stable in July. Full docs https://code.visualstudio.com/api/extension-guides/chat

We are excited to see what you build ✨ 

---
 
-- Proposal dts: https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.chatParticipant.d.ts
- Sample: https://github.com/microsoft/vscode-extension-samples/tree/main/chat-sample
- Docs: https://code.visualstudio.com/api/extension-guides/chat

Extension authors can subscribe to this issue to get updates about the proposed chat agent API. There may still be breaking changes coming, and I will post here whenever a breaking change is made.

I'm also interested in feedback about how you might use this API.

Current TODOs
- [x] Need to enable agents to identify sessions uniquely and store state, including across a window reload
- [x] Should be able to use `shouldRepopulate` for agents
  - Maybe this should be the default behavior for agents
- [x] `ChatAgentReplyFollowup` should parameterize the agent and slash command names, don't require the agent to include them in the prompt as text
  - Maybe this goes for variables too, but variables like `#file` are complex
- [x] https://github.com/microsoft/vscode/issues/204539
- [x] https://github.com/microsoft/vscode/issues/203822
- [x] Change the history context shape to be a flat list of messages, instead of request/response pairs
- [x] Understand how duplicate agent IDs should be handled
- [ ] ~~Enable agents to resolve a variable value on-demand, instead of resolving all variables before invoking the agent~~

---

- [ ] Add a `persist` flag to `ChatAgentProgressMessage` for it to stay visible with a ✔️ even when more content has been added.
  - This would be useful when you want the progress message to remain and show what happened, eg in representing a function call. I've seen multiple extension authors add their own version of this just with text.
- [ ] https://github.com/microsoft/vscode/issues/197081

---
- [x] `ChatAgentCommandFollowup` should become a progress message instead of a followup
- [x] In history, agents have to parse agent names and slash commands from the raw `ChatMessage`. They shouldn't have to understand the input syntax- the type in history should be related to the request and progress/result types that we already have.
- [x] Add generic type parameter for the type of the result, which can include custom properties
- [x] 'variables' proposal is referenced from chatAgents2, need to merge it in
- [x] https://github.com/microsoft/vscode/issues/200598

**Comment by bionicles** [comment]
any comment on this https://archive.ph/N5iVq#selection-1895.0-1897.140 ? 
![Screenshot 2023-12-06 141137](https://github.com/microsoft/vscode/assets/24532336/41a5ec3a-d8f9-4819-8c95-7013ecd4da5c)
why does github copilot think i need to talk to a lawyer to know if it's OK to use github copilot to develop AI?

**Comment by ntindle** [comment]
Has any research been done into the [agent protocol](https://github.com/AI-Engineer-Foundation/agent-protocol) for this?

**Comment by InTheCloudDan** [comment]
Do you have any example use cases of `ChatAgentCompletionItemProvider`?

**Comment by pelikhan** [comment]
- Do you have more information about the options to `access.makeRequest`?
- Do you have more information about ChatAgentUsedContext and MappedEditsProvider?

**Comment by roblourens** [comment]
@bionicles I wouldn't rely on Copilot Chat for legal advice, talking to a lawyer may be a good idea.

@ntindle I think that's a different use of the term "agent" than how we're using it here, where it's just a UX paradigm for an extension that can execute a chat query. That might continue to be a point of confusion as people work more on AI agents that plan and use tools to accomplish a task in multiple steps.

@InTheCloudDan Not a good example, sorry, I'll try to add one to the API sample this month. The API might need a little more polish.

@pelikhan The current concept is that those options could be used differently by different chat providers. The `copilot` chat provider takes a numeric option `n`. 

Re: `MappedEditsProvider`- I think this is sort of in an experimental state, there is one that is implemented by the Copilot Chat extension and just tries to fix the indentation of pasted codeblocks.

**Comment by ntindle** [comment]
How would you feel about calling it one of the following then sidekick, partner, assistant, or tool?

The name confusion I can see continuing

**Comment by bionicles** [comment]
@roblourens duh

**Comment by roblourens** [comment]
Breaking change:

Renamed `ChatVariableContext#message` to `prompt`

https://github.com/microsoft/vscode/blob/b8baf8e153c5e1194f0d7b8593950e429829789e/src/vscode-dts/vscode.proposed.chatAgents2.d.ts#L420-L425

**Comment by roblourens** [comment]
You can now return a `ChatAgentProgressMessage` in the middle of a response- previously it only worked at the top. Once there is some content to render after the progress message, it will be removed.

`ChatAgentTask` will be removed soon, in favor of just sending a progress message followed by the content.

**Comment by roblourens** [comment]
Possibly breaking change- the shape of the history context has changed from ChatMessage objects to objects that have the same request/response format as the rest of the chat agent API.

https://github.com/microsoft/vscode/pull/202541/files#diff-d005ac8b5c4b2d6ad73fbab2d3cf97636a96cf7bc11ad8339ef4d0d2dd1d15dcR18

**Comment by MRayermannMSFT** [comment]
> Possibly breaking change- the shape of the history context has changed from ChatMessage objects to objects that have the same request/response format as the rest of the chat agent API.
> 
> https://github.com/microsoft/vscode/pull/202541/files#diff-d005ac8b5c4b2d6ad73fbab2d3cf97636a96cf7bc11ad8339ef4d0d2dd1d15dcR18

@roblourens is the reason that `ChatAgentHistoryEntry` doesn't look more like the typings related to follow ups, like this:

```ts
export interface ChatAgentHistoryEntry<TResult extends ChatAgentResult2> {
	request: ChatAgentRequest;
	response: ChatAgentContentProgress[];
	result: TResult ;
}
```

Is because the history can contain entries from more than just your agent, thus the generic typing wouldn't be correct? Perhaps a typing like this could make sense?

```ts
export interface ChatAgentHistoryEntry<TResult extends ChatAgentResult2> {
	request: ChatAgentRequest;
	response: ChatAgentContentProgress[];
	result: ChatAgentResult2 | TResult;
}
```

Or if not (because I can still imagine other problems with the above suggestion), could the JSDoc on `result` make it clear if any properties added on a returned `ChatAgentResult2` will still be present in history/will be the exact same instance (similar to the JSDoc on `ChatAgentResult2Feedback.TResult`)?

**Comment by roblourens** [comment]
Yeah, but it's even a bit more complicated and maybe needs more thinking.

Before the window reload, your agent will get the same result instances that it returned. But after a window reload, when you're looking at persisted conversation history from a previous window instance, those result instances don't exist anymore, so you will only get the basic result object.

I'm trying to decide whether that's too confusing, and so we should never return the same result instance. If you can't count on it, I'm not sure it helps anyway.

**Comment by roblourens** [comment]
As I mentioned before, `ChatAgentTask` is deleted, and the progress message should be able to cover the same use case. https://github.com/microsoft/vscode/pull/202777

**Comment by jrieken** [comment]
> I'm trying to decide whether that's too confusing, and so we should never return the same result instance. If you can't count on it, I'm not sure it helps anyway.

Random alternatives that pop into my mind

* persist/restore (per extension) the JSON representation of things (don't be instance-true but data-true)
* be instance-true but for restored sessions have some flag that tells extensions that "this is restored"

**Comment by colombod** [comment]
Following a very interesting conversation with @isidorn I would like to share some thoughts I had and hopefully discuss them further.

I am finding the type `ChatAccess` confusing. Compared to other frameworks (LlamaIndex, SemanticKernel) this looks more like a context / service collection.

I can imagine this becoming even more powerful and enabling access to other agents, messaging api and other models (like embedding generator).

I think this code could be closer to what developers write in other platforms:
```js
let context = await  vscode.chat.requestContext(id);
let response = await  context.llm.makeRequest(...)l
```
This could also be the place to plug apis to get access to chat history, previous messages, and any other information that is relevant to the current chat turn.

Having access to the conversation I think is crucial as different agents could be invoked during an engagement with the chat component and those messages could be useful in creating the arguments for the `makeRequest` method.

Another thing I would like to suggest is to use a type instead of string for the request content. The OpenAI sdk (but also other implementation) are preparing for more complex conversations with multi model system. In that case messages have a content property that allows multiple parts with different mime types.

**Comment by pelikhan** [comment]
- the logic that automatically adds the missing backticks for code regions seems to be defeated when the opening code region is using more then 3 backticks (this only happens while streaming).

**Comment by roblourens** [comment]
> the sub commands are automatically surfaced at the "global" level which could lead to sub command polution. having a flag to exclude a sub command from the global context would be nice.

You can find a subCommand with `/` globally but this just inserts `@agent /command`. It should work with multiple subCommands on different agents that have the same name.

> RAG is an essential feature missing from the agent API.

We have this in mind- we may try to somehow expose some of what `@workspace` can do, like via a variable. Or access to embeddings. But no details yet.

> the chatmessagerole has "function" however it's unclear how to implement OpenAI function/tools using the current chatprovider interfaces so it's unclear what this role means. Also, in the OpenAI impl, there is a "ToolCall" vs "ToolResponse" message

We are talking to a service that doesn't support all of this yet, so only basic chat requests are supported  right now, unless you talk to OpenAI yourself.

> the used content reference messages are not part of the history

I thought it seemed that only real content messages should be part of history. Do you want the used content? Would you do something with those file entries when they appear in history?

> the logic that automatically adds the missing backticks for code regions seems to be defeated when the opening code region is using more then 3 backticks (this only happens while streaming).

Fixed, thanks

**Comment by pelikhan** [comment]
> You can find a subCommand with / globally but this just inserts @agent /command. It should work with multiple subCommands on different agents that have the same name.

It is not so much a problem of name clashes, as a problem of crowdin the dropdown with a gazilion commands.

> I thought it seemed that only real content messages should be part of history. Do you want the used content? Would you do something with those file entries when they appear in history?

If I know what content was used in previous queries, I can also add them to my context when executing the next LLM query.

**Comment by jrieken** [comment]
fyi - the progress-object that's passed into `ChatAgentHandler` will change. Instead of calling `progress.report(typeA)` and `progress.report(typeB)` there is method per display-type, like `stream.markdown(value)` and `stream.reference(value2)` etc. The new typings are here: 

https://github.com/microsoft/vscode/blob/3843819e9576681b548f3c656979d76d36f2ee75/src/vscode-dts/vscode.proposed.chatAgents2.d.ts#L282

The `report` method and the corresponding types will be removed soon

**Comment by pelikhan** [comment]
> fyi - the progress-object that's passed into `ChatAgentHandler` will change.

is it already in the nightly build?

**Comment by jrieken** [comment]
not yet, it will be tomorrows insiders

**Comment by pelikhan** [comment]
There is some weird rendering going on when one sends markdown code sections as progress messages.
<img width="525" alt="image" src="https://github.com/microsoft/vscode/assets/4175913/0e13c13d-94b6-4212-954d-cc91ddef2406">

**Comment by roblourens** [comment]
"Command followups" are gone, and have been replaced by a `command` method on the response stream, which renders the same kind of button (but now inline, and doesn't have to be at the end of the content)

https://github.com/microsoft/vscode/pull/204512/files#diff-d005ac8b5c4b2d6ad73fbab2d3cf97636a96cf7bc11ad8339ef4d0d2dd1d15dc

**Comment by jrieken** [comment]
fyi - we are going to rename `ChatAccess` and friends from the `chatRequestAccess`-proposal to "LanguageModel", e.g  `requestLanguageModelAccess`, `LanguageModelAccess`, etc. The old name was chosen because we expect a language model with chat characteristics but we understand the confusing overlap with VS Code's chat feature and chat agent. This rename also happens to make it more clear that language model access can be used independent of chat agents

**Comment by roblourens** [comment]
The command followups change above didn't actually land- trying again today

**Comment by InTheCloudDan** [comment]
How does `#file` provide a `QuickPickWindow` if the `resolve` method on https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.chatAgents2.d.ts#L578 is not async?

**Comment by jrieken** [comment]
fyi - `ChatMessageRole#Function` will be removed with https://github.com/microsoft/vscode/pull/204637

**Comment by roblourens** [comment]
Sorry, that's an internal variable that the extension API can't support yet.

**Comment by InTheCloudDan** [comment]
@roblourens got it, can we expect to see some updates here in the near future?

**Comment by roblourens** [comment]
Maybe, but no immediate plans. What are you wanting to do with it?

**Comment by InTheCloudDan** [comment]
> Maybe, but no immediate plans. What are you wanting to do with it?

I have an extension that loads data from our service into it and there's 2 top of mind resources(that each item in that resource has a unique name) that I want to make it easier for users to select and know what they and interact with Copilot around. These easiest use case I have for it is summarization but there's others I'm exploring.

Before chat variable I was timing out on `chatAccess` waiting for the user to select, I'm going to be testing it shortly but it's still not a great experience that they can select the resource but then they're not actually sure that they selected the right one, where with `#file` it is clear what resource was selected giving users confidence.

I still have more exploring to do with determining how Copilot actually includes that as context.

**Comment by jrieken** [comment]
>  I was timing out on chatAccess waiting for the user to select,

FYI - there is no more timeout

**Comment by roblourens** [comment]
I have again merged the change to switch from command followups to command buttons: https://github.com/microsoft/vscode/pull/204548 and I expect it to stick this time and be in the next Insiders build on Monday

**Comment by jrieken** [comment]
FYI - more changes to `src/vscode-dts/vscode.proposed.chatRequestAccess.d.ts`

* please update to the latest API definition for naming changes
* please note that `vscode.chat.languageModels` and `onDidChangeLanguageModels` exits. Those allow extension to know what language model (ids) actually exist
* with next copilot-chat release those are `copilot-gpt-3.5-turbo` and `copilot-gpt-4`, the old `copilot` id will _not_ work anymore

**Comment by roblourens** [comment]
A breaking change to `ChatAgentResult2`- instead of returning the same instance in places like Followups, Feedback, and history, the Result supports a `metadata` field of serializable data. You can attach whatever JSON.stringify-able data or IDs you like to the result object and look it up where Result is passed back. The result metadata will only be present on Results that came from the same agent. An advantage of this is that the state will still be available on Results after reloading the window.

https://github.com/microsoft/vscode/pull/204851/files#diff-d005ac8b5c4b2d6ad73fbab2d3cf97636a96cf7bc11ad8339ef4d0d2dd1d15dc

**Comment by roblourens** [comment]
Breaking change- the `subCommandProvider` is renamed to just `commandProvider`: https://github.com/microsoft/vscode/pull/204911

**Comment by jrieken** [comment]
please note

* `ChatMessage#name` has been removed
* `ChatAgentContext#history` will change into a sequence of request- and response-turns (also allowing for request, response, response, response, request etc)
* `ChatAgentRequest#prompt` and `ChatAgentRequest#variables` will change so that prompt is the prompt as typed (w/o agent and command part) and variables-info includes start/end within prompt

**Comment by pelikhan** [comment]
could you add a union type for ChatAgentHistoryEntry.response ?

**Comment by matv-stripe** [comment]
Hi @roblourens , thanks for working on this it seems very interesting! Are there plans to allow mentioning the agent in the context of an inline chat?

**Comment by jrieken** [comment]
Please note that with next insiders

* `LanguageModelAccess#makeChatRequest` must be called with `LanguageModelMessage` instances. The old ChatMessage interface isn't supported anymore

**Comment by roblourens** [comment]
Hey all, just wondering whether anybody has created a custom chat variable using the `chat.registerVariable` API? Would be curious to hear what you used it for and if you have feedback on that part of the API.

**Comment by lukka** [comment]
@roblourens I created a PoC of a chat variable which is provided by the C++ language service, I'll share the detail with @isidorn soon showing the scenario where it is being used. It would be nice if:
- the user could stick a variable in the prompt, in a way that does not need to be manually written over and over.
- or approaching the problem from a different angle: it would be much better if Copilot Chat would allow language services to contribute to the prompt (e.g., the `@workspace` "project labeling" phase could be one place where this could be happening, but anywhere else would be fine) so that the content of the variable is pulled into the prompt.

**Comment by jrieken** [comment]
Please note that with next insiders

* the `chatRequestAccess` proposal has been renamed to `languageModels`
* there is a new namespace `vscode.lm` for language model related functions, e.g `vscode.lm.requestLanguageModelAccess`

**Comment by roblourens** [comment]
Added `isSticky` to the chat agent, so agents can opt in to the 'sticky' behavior that slash commands have now: https://github.com/microsoft/vscode/pull/205322

**Comment by isidorn** [comment]
@roblourens should it be opt-in or opt-out?

**Comment by roblourens** [comment]
We decided to make it opt-in for now. It might be opt-out in the future.

**Comment by roblourens** [comment]
Since the term "agent" was giving the wrong impression, we're going to rename "chat agent" to "chat participant". Hopefully everyone is ok with this 😁

**Comment by greazer** [comment]
Personally, I don't love it. Trivially it's too long and it's suitable for abbreviation. But more importantly, it doesn't imply the concept behind it very well. To me a "particpant" is just a generic term for anybody that wants to be part of chat. Doesn't imply any expertise or knowledge at all. 

Maybe you've thought through this a bunch elsewhere, but what about: Aide, Ally, Broker or Rep (for representative)?

**Comment by roblourens** [comment]
> To me a "particpant" is just a generic term for anybody that wants to be part of chat. Doesn't imply any expertise or knowledge at all.

Yep, that is exactly the idea

**Comment by stepro** [comment]
FWIW, we've been using the general term "actor" to represent this "chat participant" concept. It encompasses users as well as automated entities based on the role they play in the conversation. This aligns at least with how OpenAI thinks about conversations, where each conversation entry has a role and optionally a name if there are multiple actors with that same role in the conversation. I don't think "participant" works unless you a) recognize a user is also a participant and b) have this further role categorization to identify when it's the user vs. the "other thing" (I'm currently using the term "agent" but used "assistant" in the past).

In a related note, it's worth pointing out that the foundation models from OpenAI have clearly been fine tuned to some extent using the term "assistant" (this is what they use as the role name for the AI in the underlying message delimiters). I've found the AI to get confused more often when you try to use a different term like "agent" in instructions that talk about "the user" vs. "the <insert favorite term here that isn't assistant>".

**Comment by greazer** [comment]
> We want something "free of expectations" but most importantly, this is just an API symbol name and not a product/feature name

Ok, seems reasonable. But then how are we going to refer to these "participants" from feature perspective? Still the @workspace agent and @vscode agent, @cat agent, etc?

**Comment by roblourens** [comment]
The `fullName` property moved to the "default participant" only, so it can't be set by other participants now

**Comment by caleb-at-pieces** [comment]
> * Proposal dts: https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.chatAgents2.d.ts
> * Sample: https://github.com/microsoft/vscode-extension-samples/tree/main/chat-agent-sample

These two links appear to be broken at this point.

**Comment by isidorn** [comment]
@caleb-at-pieces fixed. Thanks

**Comment by swatDong** [comment]
Is there any way to customize the order of sub commands?
For example, my agent provides some functionality commands (`create`, `update`, ...) and one `help` command, and I'd like to pin the `help` command always at the end of the list.

**Comment by WilmarE** [comment]
> Is there any way to customize the order of sub commands? For example, my agent provides some functionality commands (`create`, `update`, ...) and one `help` command, and I'd like to pin the `help` command always at the end of the list.

**Comment by isidorn** [comment]
Docs are live on our website:

https://code.visualstudio.com/api/extension-guides/chat
https://code.visualstudio.com/api/extension-guides/language-model

This is v1, we will be iteratively improving and updating them.

**Comment by lijie-lee** [comment]
Hi there, I have some questions when using the Chat Participant API.

- **limit token**: I always got "Error: Message exceeds token limit" whereas my long message is about 2k tokens (calculated by tiktoken). As far as I know, the limit token is 4K. Is the limit token dedicated by the extension? Or there are some token reversed by the ChatParticipant API self for some specific purpose like security, compliance, etc. 
  > _When I directy put the long message in the chat without any agent or command, I could get the response I need_
- **chat session lifecycle**:  I've done few test and found out
    1. After I talked to copilot chat, and then close VSCode and reopen, the chat will keep the chat history
    2. When I open a "New Chat", the history will be clear. And when I switch to previous chat, the chat history will still be there.

    So, Is the chat is alway the same one expect I create a New Chat? Is that true? and could you share more detail about that?
- **interface LanguageModelResponse**: I want my extension response a default message to user when I got error from "makeChatRequset", like exceeds limit token. But the `result` attribute of `LanguageModelResponse** is a unknow type. So could you share more dtail about that?

thanks a lot.

**Comment by vudayani-vmw** [comment]
Hi All, I have a question related to the `vscode.chat.requestLanguageModelAccess` method when interacting with the Copilot Language Model.

Issue:
The usage of the `vscode.chat.requestLanguageModelAccess` method does not seem to be working as expected for me. According to the documentation, the supported language models include `copilot-gpt-3.5-turbo` and `copilot-gpt-4`. However, when inspecting the vscode.lm.languageModels object, only the `copilot` model is listed. Subsequently, attempts to utilize this language model do not yield any response.

- Are there specific permissions required to access the copilot-gpt-3.5-turbo and copilot-gpt-4 models?
- Could you please provide additional details or insights into using these language models with the vscode.chat.requestLanguageModelAccess method?

Environment Details:
Visual Studio Code Version: 1.87.0-insider (Universal) Commit: 11095a2781bd2a14cbaddc1b925f196f1e9452bc
Chat APIs Enabled: "chatAgents2", "languageModels"

Thanks a lot.

**Comment by jrieken** [comment]
@vudayani-vmw you need to update the extension. Make sure to be on pre-release.

**Comment by jrieken** [comment]
fyi - that Language Model Access (`vscode.lm.XYZ`) is now tracked here: https://github.com/microsoft/vscode/issues/206265

**Comment by nalbion** [comment]
It would be great if GitHub apps could support similar UX:

- Currently it's not possible to `@mention` a bot user from GitHub issues or PR comments.
- GitHub's [slash commands are not available to GitHub apps](https://github.com/orgs/community/discussions/40299#discussioncomment-8611312) either.

**Comment by nalbion** [comment]
> Has any research been done into the [agent protocol](https://github.com/AI-Engineer-Foundation/agent-protocol) for this?

I've managed to get agent protocol working with this API - https://github.com/nalbion/agent-adapters/tree/main/src/agents/adapters/agentprotocol

**Comment by roblourens** [comment]
Breaking change- we have moved to a model where a static contribution in package.json is required for your chat participant and slash commands. The sample has been updated: https://github.com/microsoft/vscode-extension-samples/blob/main/chat-sample/package.json#L27-L40

This also means that your extension will not be activated until your chat participant is invoked. If you previously used the `onStartupFinished` activation event or another to ensure your extension was active before chat was opened, you should remove that.

**Comment by pelikhan** [comment]
That is a major step backward in terms of the flexibility of the APIs. Our extension populates the commands dynamically based on assets in the workspace.

**Comment by matv-stripe** [comment]
> Breaking change- we have moved to a model where a static contribution in package.json is required for your chat participant and slash commands. 

Could we get more context behind this?

**Comment by roblourens** [comment]
@pelikhan can you tell me more about your scenario? If you are enabling/disabling a fixed set of commands, the slash command contributions have a `"when"` clause with which you can enable and disable your slash commands.

> Could we get more context behind this?

In general we try to avoid having extensions activate before they are actually needed. Here, we decided that we typically expect a fixed set of chat participants and slash commands. Sorry that this is such a late change, I thought we might avoid it, but performance considerations won out. 

Since all slash commands are available before you have actually used the participant (when you type "/"), they also need to be listed in the static contribution.

**Comment by matv-stripe** [comment]
@roblourens , I am curious if it would be possible to have buttons trigger follow up chats. Right now it seems that they can only trigger commands, but it would be interesting if a `stream.button{{})` could initiate a followup (similar to the follow up provider.

**Comment by roblourens** [comment]
Can you tell me more about what you want to do?

**Comment by pelikhan** [comment]
@roblourens we expose from prompts (stored in the workspace) as commands. this is not high priority though

**Comment by roblourens** [comment]
The FollowupProvider now takes a `ChatContext`, which may make it easier to generate followups that follow from chat history

https://github.com/microsoft/vscode/blob/f838112f066283bcce562b56766b3816112ee77c/src/vscode-dts/vscode.proposed.chatParticipant.d.ts#L181

**Comment by roblourens** [comment]
Yeah, you can use `lm` and return a followup from that.

**Comment by volatile-static** [comment]
> Yeah, you can use `lm` and return a followup from that.

So what prompt do you use to generate such a followup? Will this increase the history?

**Comment by roblourens** [comment]
Whatever prompt you want to use 😁

**Comment by roblourens** [comment]
@burkeholland the variable can't cancel the chat- it can only resolve to a value. The chat participant is what's in control, and if it wants to, it could see that the variable is missing a value and it can return an error or write some other content to the response. Check out the `ChatErrorDetails` if you want to return a message rendered like an error from the participant.

@matv-stripe can you tell me more about the specific scenario? I think that writing more content to the chat after running a command isn't what I've had in mind.

**Comment by lukka** [comment]
> @burkeholland the variable can't cancel the chat- it can only resolve to a value. The chat participant is what's in control, and if it wants to, it could see that the variable is missing a value and it can return an error or write some other content to the response. Check out the `ChatErrorDetails` if you want to return a message rendered like an error from the participant.

@roblourens I am currently returning `undefined` from the `ChatVariableResolve.resolve():ChatVariableValue` when the variable is used in a context that is not appropriate and it would resolve to an empty string. I think I'd like to return a message that warnings the user about it (that is, "the variable passed in has been ignored"), and using the mentioned `ChatErrorDetails` looks good, but it cannot be used as the resolver does not return that type. Any sample or guidance would be appreciated.

**Comment by amitjoshi438** [comment]
> Breaking change- we have moved to a model where a static contribution in package.json is required for your chat participant and slash commands. The sample has been updated: https://github.com/microsoft/vscode-extension-samples/blob/main/chat-sample/package.json#L27-L40
> 
> This also means that your extension will not be activated until your chat participant is invoked. If you previously used the `onStartupFinished` activation event or another to ensure your extension was active before chat was opened, you should remove that.

I tried the sample extension but I don't see @cat coming as a chat participant. I am using insiders and on pre-release version of Github Copilot Chat.

<img width="258" alt="image" src="https://github.com/microsoft/vscode/assets/54068463/f6fbb351-e53c-45b3-91b7-b33018807eb1">

**Comment by nalbion** [comment]
> Breaking change- we have moved to a model where a static contribution in package.json is required for your chat participant and slash commands. The sample has been updated: https://github.com/microsoft/vscode-extension-samples/blob/main/chat-sample/package.json#L27-L40
> 
> This also means that your extension will not be activated until your chat participant is invoked. If you previously used the `onStartupFinished` activation event or another to ensure your extension was active before chat was opened, you should remove that.

Why? I've been working on an extension which allows user-configurable agents to be added/removed dynamically.

**Comment by nalbion** [comment]
> Since all slash commands are available before you have actually used the participant (when you type "/"), they also need to be listed in the static contribution.

I don't think it would significantly affect performance for the chat interface to accept participant additions/removals dynamically (as it used to) and update the agent slash command lists at this time

**Comment by isidorn** [comment]
@amitjoshi438 I just tried with latest vscode insiders, latest chat pre-release, and latest chat sample. And all is good. Make sure you are using the latest version of the sample

@nalbion Rob nicely answered this already here https://github.com/microsoft/vscode/issues/199908#issuecomment-1973606329 In VS Code performance is a top priority for us always, and having extension activate as late as possible is a big part of that.

**Comment by nalbion** [comment]
@roblourens I've got a few posts in this Twitter thread showing my progress and plans - https://twitter.com/nalbion/status/1757321758389288995

The user can have a config file in their home directory which my extension monitors for changes and dynamically creates/disposes/updates agents.

**Comment by andrey-justo** [comment]
Hi, I've been testing this proposal (congrats) and I wonder if you plan to:

1 - Adding arguments to ChatFollowUp to encapsulate hidden values like json in a dynamic follow up provider.
Here's an example:
```
 export const handleFollowUp= (objects: any[]): vscode.ChatFollowup[] => 
    objects.map((element: {name: string, id: string }) => {
            return {
                prompt: `Show ${element.name}`,
                label: vscode.l10n.t(`Show ${element.name}`),
                command: "/showCommand",
                arguments: [element.id] // -- new option
            } satisfies vscode.ChatFollowup
        });
            .
            .
            .
export const handleCommand= 
    async (request: vscode.ChatRequest, context: vscode.ChatContext, stream: vscode.ChatResponseStream, token: vscode.CancellationToken): Promise<IChatResult> => {
        stream.markdown(`You selected ${request.arguments.id}`);
    };
```

2 - Do you plan to let chat button (interpreted as commands) to have capabilities to send chat questions when the user interacts with buttons inside of it?
Example:
```
export const handleCommand= 
    async (request: vscode.ChatRequest, context: vscode.ChatContext, stream: vscode.ChatResponseStream, token: vscode.CancellationToken): Promise<IChatResult> => {
         stream.buttn({
            title: `Show helloworld`,
            command: "/helloWorld",
            arguments: ["javascript"]
        })
    };
    
export const handleButton= async (argument: string,  stream: vscode.ChatResponseStream) => {
    stream.markdown(`
          // Hello World Program in {argument}
         console.log('Hello World');
    `);
}
...

// extensions.ts file
...
context.subscriptions.push(participant, vscode.commands.registerChatCommad("/helloWorld" handleButton));

```

**Comment by matv-stripe** [comment]
> 2 - Do you plan to let chat button (interpreted as commands) to have capabilities to send chat questions when the user interacts with buttons inside of it?

I think this would be a nice feature!

**Comment by roblourens** [comment]
@andrey-justo 

1. No, but if there is some hidden state you need, you can return a `ChatResult` object from your handler with arbitrary data in its `metadata` field, and reference that in the history on the next message.
2. You can use the `workbench.action.chat.open` command for this with arguments described here: https://github.com/microsoft/vscode/blob/868cac2319b56115da363b818b7167f87274bb9f/src/vs/workbench/contrib/chat/browser/actions/chatActions.ts#L42-L52

**Comment by formulahendry** [comment]
Has Chat API been finalized now? And will the Chat and LLM API be stable API in v1.88?

**Comment by roblourens** [comment]
No, it won't be finalized this month

**Comment by lijie-lee** [comment]
Will the Chat Participant API expose token number related interface?
In that way, our participant could 
- control the request prompt not exceed the limitation. 
- obtain the request / response token number for purpose, like, improving performance.

As far as I know, there are some packages for tokenizer, like [tiktoken](https://www.npmjs.com/package/tiktoken), but not know if the count calcuated by them are same as the Chat API's.

**Comment by InTheCloudDan** [comment]
Are there plans to add `icon` support to the markdown processing in the future? Using the `$(alert)` syntax? I just tried it calling `stream.markdown(`$(alert)`);` to confirm it didn't work.

**Comment by InTheCloudDan** [comment]
did `ChatAgentCompletionItemProvider` end up getting removed? I'm looking to have a variable that is only invokable in the context of our participant.

**Comment by randi274** [comment]
I understand that building an extension using the Chat Participant API today requires a [dependency](https://code.visualstudio.com/api/extension-guides/chat#develop-a-chat-extension) on the copilot chat extension. Does this also require end users of extensions built with this dependency to have copilot licenses? Is this the long-term plan for working with the chat participant API? And are there other proposed APIs being built with this dependency in mind?

**Comment by foreverCP3** [comment]
Is there any best practice that could help me adapt to compatibility issues caused by VSCode upgrades?

从 Windows 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>发送

________________________________
发件人: Randi Wilson ***@***.***>
发送时间: Saturday, March 16, 2024 2:05:16 AM
收件人: microsoft/vscode ***@***.***>
抄送: Jianhua Ren ***@***.***>; Manual ***@***.***>
主题: Re: [microsoft/vscode] Chat participant API (Issue #199908)


I understand that building an extension using the Chat Participant API today requires a dependency<https://code.visualstudio.com/api/extension-guides/chat#develop-a-chat-extension> on the copilot chat extension. Does this also require end users of extensions built with this dependency to have copilot licenses? Is this the long-term plan for working with the chat participant API? And are there other proposed APIs being built with this dependency in mind?

―

**Comment by InTheCloudDan** [comment]
@jrieken thanks, I missed that 2nd parameter, but even with it set, it is not working
```
			stream.markdown(new vscode.MarkdownString(`$(mark-github)`, true));
			stream.markdown(new vscode.MarkdownString(`$(alert)`, true));
```

![CleanShot 2024-03-18 at 09 25 49](https://github.com/microsoft/vscode/assets/10308834/d8caef02-5f5b-4c9c-bf8b-3d4c26c34b83)

**Comment by roblourens** [comment]
I think only the `enabledCommands` field works on MarkdownString right now in the chat response, but we should support all of the fields. I opened https://github.com/microsoft/vscode/issues/208096 for this.

**Comment by roblourens** [comment]
> did ChatAgentCompletionItemProvider end up getting removed? I'm looking to have a variable that is only invokable in the context of our participant.

No, but it's in the chatParticipantAdditions proposal and won't be finalized with the first round.

> Does this also require end users of extensions built with this dependency to have copilot licenses?

This is the plan for now

> Is there any best practice that could help me adapt to compatibility issues caused by VSCode upgrades?

Sorry that this API has seen some churn, but that is typically the case when proposed APIs are under development. When the API is finalized, it will be stable.

**Comment by matv-stripe** [comment]
Is it possible to create an input field in the chat window? For context, we are exploring generating some code, returning it in the chat window, but we want the user to be able to edit some placeholders in that code. If an input field is not possible, are there other alternatives?

**Comment by roblourens** [comment]
@matv-stripe If you could edit the code block in the response, would that do it for you?

**Comment by roblourens** [comment]
Breaking change! Chat participants will be registered with an "id" and a "name". The name is what the user will use to invoke the participant with `@`. Since we can't prevent extension authors from create chat participants with the same name, this will enable us to support multiple participants with the same name.

Here's an example of how this is adopted in the chat sample: https://github.com/microsoft/vscode-extension-samples/pull/992

The VS Code PR: https://github.com/microsoft/vscode/pull/208142

**Comment by stepro** [comment]
General question. Each month, some set of breaking API changes make their way from insiders to stable. Is there a way to get an aggregate of all these changes and ahead of time (ideally a week or so) so that we have a chance to update an extension we've shipped using the proposed APIs?

**Comment by roblourens** [comment]
Sorry for the churn. It should be over soon. But there's really no way to give that kind of advance notice with the tight schedule that we are working on.

**Comment by stepro** [comment]
I mean any amount of advance notice would be nice :) Just so I understand, at the start of each month, whatever is in insiders becomes what's in stable? So I could look the day before (March 30?) and prepare to ship a new version of my extension on April 1 based on what I currently see checked in for proposed APIs and how that differs to the snapshot I have checked in for my extension?

**Comment by varughese** [comment]
> @matv-stripe If you could edit the code block in the response, would that do it for you?

Yup, I think that would work!

**Comment by InTheCloudDan** [comment]
@isidorn are we allowed to use the Copilot contributed themicon if we have a command that is invoking Copilot? Or is there a more generic one we're required to use?

**Comment by MRayermannMSFT** [comment]
> 🚨 Breaking change! Chat participants will be registered with a unique id and a "name". The name is what the user will use to invoke the participant with `@`. Since we can't prevent extension authors from create chat participants with the same name, this will enable us to support multiple participants with the same name.
> 
> Here's an example of how this is adopted in the chat sample: [microsoft/vscode-extension-samples#992](https://github.com/microsoft/vscode-extension-samples/pull/992)
> 
> The VS Code PR: #208142

@roblourens we adopted the breaking changes this morning, but couldn't get our agent successfully invoked. It looked like an activation problem, so we re-added `onStartupFinished` as an activation event to our extension, and now the agent can be successfully invoked.

Can y'all double check that the breaking change didn't cause an an issue in the "activate an agent's extension when the agent is invoked" logic?

**Comment by stepro** [comment]
Minor bug I found while picking up this month's changes. The sticky bit doesn't take effect if it's also loading the extension as part of the response. Works after that though.

**Comment by roblourens** [comment]
> I mean any amount of advance notice would be nice :) Just so I understand, at the start of each month, whatever is in insiders becomes what's in stable? So I could look the day before (March 30?) and prepare to ship a new version of my extension on April 1 based on what I currently see checked in for proposed APIs and how that differs to the snapshot I have checked in for my extension?

Yes, we ship each month, you can track the schedule in our endgame plans or iteration plan issues: https://github.com/microsoft/vscode/issues/208302. So if you want your extension in VS Code stable to be stable, you can check the changes in Insiders at the end of the month.

>> @matv-stripe If you could edit the code block in the response, would that do it for you?
>
> Yup, I think that would work!

Maybe codeblocks should just always be editable in chat responses, I don't know if there's a reason not to, I will think about that for next month.

@MRayermannMSFT I think I just fixed that, apologies, please try in Monday's Insiders build

Good point re: isSticky. I think the most sensible way to deal with this is to make `isSticky` part of the static registration, same as it is for commands. That change will be in the next Insiders. Here's what the adoption looks like in the sample: https://github.com/microsoft/vscode-extension-samples/pull/993

**Comment by bwateratmsft** [comment]
@roblourens I found an issue with followups directed at an agent. Here I am testing with a new `@azure` agent. It suggests some followups, but when I choose one, it is putting the participant ID at the beginning of the message, instead of the participant name--which does not work. Screenshots explain better:

![image](https://github.com/microsoft/vscode/assets/36966225/2391f98d-adb8-4c3c-adbb-499aaf3038fe)

[Choose a followup]

![image](https://github.com/microsoft/vscode/assets/36966225/32c79e70-c95a-4379-8863-bafa36f40a1d)

Here, `@ms-azuretools.azure-agent` is prepended instead of `@azure`. Consequently, the `@azure` agent is not invoked.

**Comment by roblourens** [comment]
I think I also fixed that on Friday, are you sure you've updated to the latest VS Code?

**Comment by bwateratmsft** [comment]
Yup:

Version: 1.88.0-insider (user setup)
Commit: b0d975fc6370f15570c380f41015b1ca2cdcb791
Date: 2024-03-25T05:49:32.344Z
Electron: 28.2.6
ElectronBuildId: 27476517
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Windows_NT x64 10.0.22631

**Comment by roblourens** [comment]
Oops, I see it, thanks

**Comment by stepro** [comment]
> Good point re: isSticky. I think the most sensible way to deal with this is to make `isSticky` part of the static registration, same as it is for commands. That change will be in the next Insiders. Here's what the adoption looks like in the sample: [microsoft/vscode-extension-samples#993](https://github.com/microsoft/vscode-extension-samples/pull/993)

Is there any sense in considering the sticky bit actually *less* static than it currently is? For instance, could ChatResult have an `isSticky` property that indicates for this result only, it should be sticky for the next user input? This would ensure participants are very deliberate in their choice to be sticky but also allow them to be more dynamic in the decision. Say a participant knows that it's currently working through something with the user (so stay sticky) until a certain point when it knows it's done (so not sticky anymore).

**Comment by formulahendry** [comment]
In https://github.com/microsoft/vscode/issues/208302, I see 03/22 is Code freeze for the endgame.
Does it mean there will not be any breaking changes for March release (v1.88) ?

**Comment by InTheCloudDan** [comment]
Is it no longer possible to render a button in a followup? In earlier iterations I was able to, but looking at the [signature](https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.chatParticipant.d.ts#L149-L170) now and basic testing the `command` is only a Chat Command, so I can't even invoke direct VSCode extension commands with it.

**Comment by stepro** [comment]
What's the reason for disallowing a follow up from using a participant that was contributed by a different extension? I'd like to look up some other extensions that I know contribute their own participants and then based on what the user is asking my participant, I might suggest that they follow up with a question to one of these other participants I'm aware of.

I would understand the concern if one participant was able to auto-invoke another one not owned by the same extension, but the user needs to decide to take the follow up here. Does the follow up show the @mention prepended? If not, maybe that would be necessary to show it's targeting a different participant.

**Comment by Alive-Fish** [comment]
Hi @roblourens, we have also met the issue when using follow-ups. Here, @ms-teams-vscode-extension.teams is prepended instead of @teams. Consequently, the @teams agent is not invoked.

Insider version:
Version: 1.88.0-insider (user setup)
Commit: 5d8491d7130d25aae121eeaa1633603ef8faa39e
Date: 2024-03-26T05:49:49.370Z
Electron: 28.2.6
ElectronBuildId: 27476517
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Windows_NT x64 10.0.22631

Does the usage of follow-ups change or is it an issue you will fix?

![image](https://github.com/microsoft/vscode/assets/15262146/3e67bdb9-160a-421d-a241-88a6238ea35c)

**Comment by lukka** [comment]
@roblourens @jrieken @isidorn when a chat variable is being mentioned, the context should come from the active document where the selection is currently living in. I am currently using `vscode.window.activeTextEditor?.document` to fetch the context, and I wonder if it would be beneficial to have that expressed in the [ChatVariableContext ](https://github.com/microsoft/vscode/blob/191be39e5ac872e03f9d79cc859d9917f40ad935/src/vscode-dts/vscode.proposed.chatVariableResolver.d.ts#L37) instead.

@hnrqbaggio @benmcmorran @spebl FYI

**Comment by ashrafulsust** [comment]
Hi @roblourens, Is there a current timeline for when a stable version of the proposed API might be released?

**Comment by bentruyman** [comment]
I didn't see this [`vscode.lm.registerLanguageModel`](https://github.com/microsoft/vscode/blob/b4378dfa1d3c37f1b38b1f897fdf33dc01e445d9/src/vs/workbench/api/common/extHostLanguageModels.ts#L143) function referenced in the `d.ts` files related to this feature — but does this function mean there can be language model providers that aren't OpenAI? Like we could register an Ollama-based model provider? Complete with its own auth provider, prompt templating, etc.?

**Comment by jrieken** [comment]
@bentruyman That's in the `src/vscode-dts/vscode.proposed.chatProvider.d.ts` proposal which has matured and isn't scheduled for finalisation

**Comment by EronWright** [comment]
I wrote an experimental chat participant that generates code based on the conversation.  I'm interested how my participant would infer which programming language to use based on the workspace codebase.

In concept, my agent could ask the workspace agent, e.g. "@workspace which language is primarily being used by the #codebase? One-word response please."  But direct agent-to-agent communication isn't supported, correct? For security reasons I assume.

The user could tell my agent, e.g. "@myagent select a language based on the #codebase".  I don't know what the resolved value of `#codebase` is or how I'd figure out the language, though the workspace agent seems able to.

I could propose a follow-up question, "@workspace which language is primarily being used by the #codebase?", but follow-ups cannot target a different agent.

As a last resort, I could try to infer the language from the conversation history, and suggest the user ask the workspace for advise (sort of what a follow-up is for).

Thoughts?

**Comment by roblourens** [comment]
@lukka anything in vscode could be used for context for a participant, so you should use the extension API to get whatever context you're interested in, like you are with `activeTextEditor`

@EronWright most likely you'd also just want to look at the language of the current editor? You could also instruct the user to use `#codebase` but I wouldn't really recommend that approach, it's not really what that variable is optimized for and is much slower than needed. You can also just look at the contents of the workspace yourself, seems faster.

@ashrafulsust No promises but I would like to finalize the chat participants API in May. And I want to point out to everyone if it isn't clear that chat variables probably won't be part of the first round of finalization. We want to keep exploring how these should work alongside things like tool calling, contributing tools from other extensions, etc, but since those scenarios are very useful, I'd expect to prioritize that soon after finalizing participants.

**Comment by nalbion** [comment]
I don't suppose the user's GutHub access token is available to extension agents?

**Comment by roblourens** [comment]
Yes, an extension is allowed to ask for it https://code.visualstudio.com/api/references/vscode-api#authentication

**Comment by MRayermannMSFT** [comment]
@roblourens it would be great if the copilot/copilot chat extension could contribute a context key which lets other extensions understand if the user is able to use copilot/chopilot chat. We're experimenting with menu items like this:

![image](https://github.com/microsoft/vscode/assets/21112697/a4681986-7aa6-4a41-a27e-8d8057d8f1b0)

And it'd be great if we could easily show/hide it based on if the user has access to copilot/copilot chat.

**Comment by roblourens** [comment]
Yes, we sort of had this but I just renamed it to `chatIsEnabled` so it makes more sense externally. Try it in the next Insiders https://github.com/microsoft/vscode/pull/210126/files#diff-9958a6665f9145da52465a4d274e59d9c9fc63398078fdada26c7fb90ca2974eR24

**Comment by MRayermannMSFT** [comment]
Hmmm as of today, trying to call `vscode.lm.sendChatRequest` results in an error from chat extension:
```
[extension] Error on conversation request: TypeError: e.processResponseFromChatEndpoint is not a function
    at Fwe (c:\Users\username\.vscode-insiders\extensions\github.copilot-chat-0.15.2024041202\dist\extension.js:1099:5751)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at Vg.fetchMany (c:\Users\username\.vscode-insiders\extensions\github.copilot-chat-0.15.2024041202\dist\extension.js:1099:10361)
    at L2.provideLanguageModelResponse2 (c:\Users\username\.vscode-insiders\extensions\github.copilot-chat-0.15.2024041202\dist\extension.js:980:1801)
```

Is this a known problem?

**Comment by isidorn** [comment]
@MRayermannMSFT sorry about that. I can also reproduce the issue. We are looking into it.

As a workaround until we release the fix, you can use VS Code Insiders from 5 days ago.
VS Code Build: 7eba2d200076e9ca04ef1d0874fb49aae426f895

darwin: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCode-darwin.zip
darwin-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCode-darwin-arm64.zip
darwin-universal: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCode-darwin-universal.zip
linux-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insider-arm64-1712728194.tar.gz
linux-armhf: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insider-armhf-1712728355.tar.gz
linux-deb-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insiders_1.89.0-1712728251_arm64.deb
linux-deb-armhf: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insiders_1.89.0-1712728408_armhf.deb
linux-deb-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insiders_1.89.0-1712729740_amd64.deb
linux-rpm-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insiders-1.89.0-1712728348.el8.aarch64.rpm
linux-rpm-armhf: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insiders-1.89.0-1712728484.el8.armv7hl.rpm
linux-rpm-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insiders-1.89.0-1712729833.el8.x86_64.rpm
linux-snap-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insider-x64-1712731301.snap
linux-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/code-insider-x64-1712728214.tar.gz
win32-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCodeSetup-arm64-1.89.0-insider.exe
win32-arm64-archive: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCode-win32-arm64-1.89.0-insider.zip
win32-arm64-user: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCodeUserSetup-arm64-1.89.0-insider.exe
win32-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCodeSetup-x64-1.89.0-insider.exe
win32-x64-archive: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCode-win32-x64-1.89.0-insider.zip
win32-x64-user: https://vscode.download.prss.microsoft.com/dbazure/download/insider/7eba2d200076e9ca04ef1d0874fb49aae426f895/VSCodeUserSetup-x64-1.89.0-insider.exe

**Comment by roblourens** [comment]
Will be fixed in the next Copilot Chat pre-release

**Comment by caleb-at-pieces** [comment]
Is there plans for this API to no longer depend on the GitHub copilot chat extension / a GitHub copilot subscription at some point?

Particularly for the case where the extension authors will not require any LLM service provided by GitHub?

**Comment by gabrielBusuioc** [comment]
@caleb-at-pieces I was asking myself the same question but it looks like all the interactive-related (or better chat-related) API is being marked as deprecated or removed(see [this](https://github.com/microsoft/vscode/commit/ea3d0c410f6a3959d7c8812318fdc84b9205f6d8) commit). Please correct me if I'm wrong...

Even though I would love to, I don`t think VS Code will provide API to develop standalone chat extensions anytime soon :(

**Comment by isidorn** [comment]
> Is there plans for this API to no longer depend on the GitHub copilot chat extension / a GitHub copilot subscription at some point?

This is not planned for the next 6 months. Right now we are in the phase of stabilising the chat participant API, language model API and variable resolver API. Having said that, the API is architected in a way that it is not tied to the Copilot Chat extension. Allowing any extension to be a default participant might be something we decide to do in the future - but this has not yet been discussed. I can share more information in the future. Thank you 🙏

**Comment by amitjoshi438** [comment]
Hi  @roblourens  I am working on a chat participant and will be calling a different LLM service to get the response. When I am trying stream.markdown() the code blocks are not displayed properly. I even tried the sample from [chat doc](https://code.visualstudio.com/api/extension-guides/chat#develop-a-chat-extension) but that is also not displayed as a code block as shared in doc.

From Doc:
![image](https://github.com/microsoft/vscode/assets/54068463/174b559e-89b8-422f-9de5-2946aee8c502)

My Code:
![image](https://github.com/microsoft/vscode/assets/54068463/7df4e4fe-0148-4201-b6d5-f88f3c2c1532)

Rendered output:
![image](https://github.com/microsoft/vscode/assets/54068463/8ebba4a6-9054-497b-b256-add55c8f423d)

**Comment by abasu0713** [comment]
I am sorry about this amateurish question.. But from what I understand from 2 days of reading documentation.. I cannot use the Chat API until I use Github's Copilot? Would that assumption be correct? 
What if someone wants to use the Chat API + A custom LSP Server to back ->  let's say completions, infilling, and additional voice commands from LLMs that are running locally with something like `ollama` or `llama-cpp-python[server]`, or HuggingFace's TGI?
I see there are extensions like: `llm-vscode` from HuggingFace that packs it's own custom `llm-ls` but it's highly unstable. It's trying to cover too many things with all the customization capabilities to support a broad spectrum of model runners. and hence not tested very well. 
And Github Copilot is an expensive service in my opinion. Sure it's great with it's Chat Controller and extensively developed features.. But in an age where powerful models are being readily open-sourced -> and powerful GPUs are readily available at home and/or ephemeral accelerated compute runtimes shouldn't the VS Code's Chat API be a little bit more open. 
I would really appreciate if someone could shed a little more light on what I might be missing and provide any suggestions.

**Comment by abasu0713** [comment]
> > Is there plans for this API to no longer depend on the GitHub copilot chat extension / a GitHub copilot subscription at some point?
> 
> This is not planned for the next 6 months. Right now we are in the phase of stabilising the chat participant API, language model API and variable resolver API. Having said that, the API is architected in a way that it is not tied to the Copilot Chat extension. Allowing any extension to be a default participant might be something we decide to do in the future - but this has not yet been discussed. I can share more information in the future. Thank you 🙏

+1 on this please.

**Comment by InTheCloudDan** [comment]
Is the Chat Participant API expected to be GA for 1.89 release?

**Comment by roblourens** [comment]
@amitjoshi438 this is working fine for me. Make sure you've got it exactly right and aren't missing a backtick. You can share a repo or a branch of the chat-sample extension if you'd like me to look at it more.

@abasu0713 see Isidor's comment https://github.com/microsoft/vscode/issues/199908#issuecomment-2074438549

@InTheCloudDan current plan is that it will be finalized in May for the 1.90 release.

**Comment by roblourens** [comment]
One more change that I want to mention here- I expect to finalize the Participant API in 1.90, but the first cut will likely be finalized without support for variables/variableResolvers. Sorry that this part will take a little longer.

The reason is that we are also working on API to enable function calling, ie "tools" or "skills". To me there appears to be a connection between "including a variable" and "calling a function" in evaluating a request. I've never been totally happy with the way variables work now, and I want to be able to build the variables API alongside skills. But I think there are important scenarios there, so I hope that it can follow quickly behind the main Participants API.

We take VS Code's extension API really seriously, and this is probably the most complex API with the most unknowns that we've had to design yet. Thanks for everyone's patience and also for the feedback I've gotten on it! Would love to keep hearing about your experiences in trying to build something cool with this API, positive or negative.

**Comment by abasu0713** [comment]
> @amitjoshi438 this is working fine for me. Make sure you've got it exactly right and aren't missing a backtick. You can share a repo or a branch of the chat-sample extension if you'd like me to look at it more.
> 
> @abasu0713 see Isidor's comment [#199908 (comment)](https://github.com/microsoft/vscode/issues/199908#issuecomment-2074438549)
> 
> @InTheCloudDan current plan is that it will be finalized in May for the 1.90 release.

thanks for the reply. I read through this entire thread only after my premature comment. :) I really appreciate it. 

It's great to know that the Language Model API is not tightly coupled with Chat API. With that being said, I am seeing the `llm-vscode` extension is using Language Model APIs (from the client side on the extension) to make API calls to it's custom `llm-ls` server. And I do see that the proposed API has landed on `master` branch. So would my assumption be correct that we could use the Language Model API with let's say a custom command (on a LSP server built with something like Rust) to fulfill let's say completions without having to use the Chat API?

**Comment by isidorn** [comment]
@abasu0713 says:
> So would my assumption be correct that we could use the Language Model API with let's say a custom command (on a LSP server built with something like Rust) to fulfill let's say completions without having to use the Chat API?

This is correct. Chat and Language Model APIs are orthogonal. I suggest to check out these guidelines: https://code.visualstudio.com/api/extension-guides/language-model

**Comment by abasu0713** [comment]
Thank you for the confirmation. This is exciting

**Comment by roblourens** [comment]
Thanks @amitjoshi438. You just need a newline before the codeblock in this case. `stream.markdown(`\n\`\`\`typescript //...`.

`stream.markdown('a'); stream.markdown('b');` is the same as `stream.markdown('ab');`. So need to make sure your codeblock lands on its own line so it is proper markdown.

**Comment by ashrafulsust** [comment]
Hi @roblourens and team, Thank you for the continued hard work on the project. I have a question regarding dynamic chat variables, specifically those like #file.

I'm interested in exploring the possibility of defining similar context variables that would require user input. Is there currently a way to achieve this functionality?

**Comment by roblourens** [comment]
No, there's nothing for that right now. I'd like to explore API for that in June or July though.

**Comment by doggy8088** [comment]
The `ChatResponseStream` seems not working as expected. All my prompts are always response so fast after wait a while time and streaming back to my editor. I was expected to get instant response (1 ~ 2 seconds) and get few words at a time in the streaming mode. I tested using this [chat-sample](https://github.com/microsoft/vscode-extension-samples/tree/main/chat-sample).

**Comment by amitjoshi438** [comment]
Hi @roblourens and team, not sure if this is already shared but is there any plan to add `when` clause for `chatParticipants` to control the participant is visibility?

**Comment by matv-stripe** [comment]
Thanks for all the hard work on shaping this API. We announced our Stripe VS Code extension with Github Copilot in private beta at Stripe Sessions last week and received positive feedback and excitement. 

Developer Keynote video snippet: https://youtu.be/lzMqzYea-3I?t=771
Main Keynote video snippet: https://www.youtube.com/watch?v=j_esVVI4KlI&t=3461s

I am particularly excited by the future work to allowing calling participants from the inline editor and richer support for variables. Not being dependent on a GitHub subscriptions would be great as well, to allow more users to have access.

It would also be great to update the button API as well such that it could pass context into an command. For example, if we could stream two buttons and have them trigger the same command, but pass an argument into the command, that would allow us for richer interactions.

**Comment by isidorn** [comment]
@AriMKatz this sounds similar to https://github.com/microsoft/vscode/issues/205874

@doggy8088 this should just work. Please file a new issue with concise reproducible steps, and feel free to ping me @isidorn on the issue

@amitjoshi438 this is not currently on the plan afaik. Can you please file a new feature request and explain your scenario and how would you use the context key

@matv-stripe I liked the presentation, congrats! For button arguments - can you file a new feature request? I do not think this was brought up yet.

**Comment by jrieken** [comment]
@AriMKatz The plan is to have both: (1) a generic API to expose/use embedding endpoints for extensions and (2) allow to reuse the logic that powers `#codebase` like as a "tool" that extensions/participants can invoke. 

> @matv-stripe I liked the presentation, congrats! For button arguments - can you file a new feature request? I do not think this was brought up yet.

Maybe I am on a different page but you do the following: `stream.button({title: 'Here', command: 'myCommandId', arguments: [1, true, {obj: isJSONSafe}]})`

**Comment by jrieken** [comment]
same thing

**Comment by jrieken** [comment]
@MRayermannMSFT we have changed things (back) to be just `LanguageModelChatMessageRole` and a single `LanguageModelChatMessage`. The `2` variant was a stepping stone and shouldn't be used anymore (will be gone tomorrow). Also, already more preemptive excuses for more changes that are coming...

**Comment by bwateratmsft** [comment]
@roblourens @jrieken what do `markdownWithVulnerabilities` and `detectedParticipant` on `vscode.ChatResponseStream` do?

**Comment by jrieken** [comment]
please check https://github.com/microsoft/vscode/issues/206265#issuecomment-2100948343 for changes in the `languageModels`-proposal, aka `sendChatRequest`

**Comment by roblourens** [comment]
@bwateratmsft Those APIs, and everything in `chatParticipantAdditions.d.ts`, aren't part of the set that will be finalized, and might not work for your extension.

**Comment by roblourens** [comment]
Scrolling way back... @amitjoshi438, maybe I should have a 'when' clause for participants. What do you want to use it for?

**Comment by amitjoshi438** [comment]
Hi @roblourens the requirements are similar to how we use 'when' clause for other contributions:

1. Showing the participant only when the opened project folder is of certain type.
2. Keeping it behind a configuration flag to allow enabling/disabling it.

**Comment by roblourens** [comment]
Yeah, that makes sense. I'll add that

**Comment by MRayermannMSFT** [comment]
A low priority issue I noticed with the chat participant experience. For a restored session, doing a thumbs down on a restored response does not activate the extension that owns that participant. Doing so would give the extension a chance to register their `onDidReceiveFeedback` event. Or in other words, I would expect an algorith like this:

1. VS Code opened
2. Chat conversation restored
3. User gives feedback on a reponse from `@SomeParticipant`
4. The extension that contributes `@SomeParticipant` is activated
5. The extension creates its chat participant, and registers a listener for the participants `onDidReceiveFeedback` event
6. Giving of feedback now happens, event is fired, extension gets to do what it wishes with the feedback

**Comment by pelikhan** [comment]
The selectlanguagemodel api locks up when using in offline or without internet. I was on the captive wifi of a plane trip and it would never return.

**Comment by jrieken** [comment]
It is `sendRequest` but it is _not_ on `vscode.lm` but on a language model chat object that you need to ask for, see https://github.com/microsoft/vscode/issues/206265#issuecomment-2100948343

**Comment by roblourens** [comment]
Wanted to call out a change that I recommend adopting- you can now set a `fullName` on your participant, which is shown in the response header. Note `@vscode` vs `VS Code` here:

<img width="223" alt="image" src="https://github.com/microsoft/vscode/assets/323878/b62e466b-b496-438d-94c7-8150689e3664">

Also, I mentioned changes to variables earlier. Currently, chat participants will be able to take the existing variables, but not contribute their own. The `variables` property on the request has been renamed to `references`, and there are some other minor changes to the shape of the reference interface. The `#codebase` variable won't show up for third-party participants, but this is only because we think that it isn't really the right API for this sort of data. I'd like to make it available via a different API soon.

All that said, we think we are finally almost ready to finalize this API! We are hoping to do some final polishing and finalize it this week. Once that happens, you will be able to publish extensions and use them in VS Code Insiders until the next stable release.

**Comment by rvanider** [comment]
@roblourens Happy to see this release getting close - great to see this functionality.

I have been trying to use `ChatParticipant.participantVariableProvider` to provide completion and scoped variable references. The chat UI presents the items, allows them to be selected and they *appear* as if they are variables/references but when the request is passed along to the handler the references property does not contain the references/variables.

I have found the `ChatVariableValue` to be sensitive to quietly working or not working depending on how it is populated.

Is there some guidance on an implementation of this part of the API that can be referenced?

**Comment by roblourens** [comment]
Only the `chatParticipants` proposal is being finalized- things in `chatParticipantsAdditions` like `participantVariableProvider` aren't being finalized yet. We use that API for some special cases in Copilot Chat, but it might not be stable enough for general use.

**Comment by matv-stripe** [comment]
> you can now set a fullName on your participant

Where do we set this?

**Comment by jamalc** [comment]
Should metadata returned by the chat request handler be available in the history on subsequent requests? I'm adding a text field to the chat result metadata to use for multi-turn chat conversations that might include non-text content.

**Comment by roblourens** [comment]
Yeah, anything on that field has to be JSON-stringifiable, but it should be available to look up on the next turn

**Comment by matv-stripe** [comment]
@isidorn when will this API be released in vscode stable?

**Comment by isidorn** [comment]
@matv-stripe we decided that the API will remain finalised only in Insiders in May and June. So start of July it will be released in stable.

**Comment by roblourens** [comment]
I hadn't mentioned it here yet- 

The chatParticipant and languageModels APIs are now finalized in today's Insiders! You can now publish your extensions to the marketplace and they will be usable in Insiders. Please let me know if you run into any issues.

As Isidor says, the current plan is that chat participants won't appear in the next Stable VS Code in June, but will later.

**Comment by InTheCloudDan** [comment]
I'm not sure where to post this issue. It just started this morning as far as I know, but I haven't been testing as much this week. It happens with any `@mention` of my participant. I tried switching to release version but that throws the type error where `sendRequest` isn't defined.

```
[Extension Host] TypeError: Cannot read properties of undefined (reading 'Symbol(Symbol.asyncIterator)')
    at n.CopilotProvider.handler [as y] (<extension>/src/providers/copilot.ts:94:46)
    at i.$invokeAgent (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:153:45525) (at console.<anonymous> (/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:143:59377))
```
    
    
```
Version: 1.90.0-insider (Universal)
Commit: a48f464a3e01aad384703ec964018299b14bb7cf
Date: 2024-05-17T05:50:01.831Z
Electron: 29.3.1
ElectronBuildId: 9464424
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Darwin arm64 23.4.0
```

**Comment by InTheCloudDan** [comment]
thanks @bwateratmsft!

**Comment by isidorn** [comment]
@InTheCloudDan good catch. I just fixed it. Thanks

**Comment by isidorn** [comment]
As Rob said we have finalized the Language Model and Chat Participant API 

Thus, you can now publish your extensions to the VS Marketplace. This API is finalized in VS Code Insiders, and will be finalized in VS Code Stable in July. Thus we suggest that you use an engines: ^1.90.0 property in your package.json, and VS Code Stable will gracefully handle your extensions and ignore your usage of Language Model and Chat API until the API gets finalized in Stable in July. 

Full docs https://code.visualstudio.com/api/extension-guides/chat

We are excited to see what you build ✨ 

Thus I am going ahead and closing this issue. We will create follow-up issues for API that is still in proposed state. Here's the issue for Variables

**Comment by roblourens** [comment]
Two more issues to follow for API updates. I'm very interested to hear what you would want to use custom chat variables or tools for, so please let me know!

- https://github.com/microsoft/vscode/issues/212995
- https://github.com/microsoft/vscode/issues/213274

**Comment by undesicimo** [comment]
Congratulations! been watching this for months so thrilled

**Comment by jasonwang82** [comment]
Technically, is it possible to turn off the chat on the left sider bar but inline chat is still there in active. @roblourens

**Comment by isidorn** [comment]
@jasonwang82 yes. But, this is unrelated to chat participant API.

**Comment by matv-stripe** [comment]
Is it possible to stream results in our Markdown that include a dropdown? 

We want to add a "References" collapsable section that is closed by default, similar to how Copilot shows "Used 1 reference >" here

![CleanShot 2024-05-23 at 13 44 21@2x](https://github.com/microsoft/vscode/assets/55463370/be6348b1-777a-463d-b80e-ba77b1b9704a)

I dont seem to see that option in the code here https://github.com/microsoft/vscode/blob/101ea2a13ea896a99a6375dacc71addcaa9a080a/src/vs/workbench/contrib/chat/browser/chatListRenderer.ts#L594

**Comment by roblourens** [comment]
You can call `reference` on the response stream to add to that list, but you can't add your own.

**Comment by amitjoshi438** [comment]
> Yeah, that makes sense. I'll add that

Hi @roblourens is there any tracking link for this?

**Comment by roblourens** [comment]
https://github.com/microsoft/vscode/issues/214379
